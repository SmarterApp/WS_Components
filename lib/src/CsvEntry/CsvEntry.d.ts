import * as React from "react";
import { ItemRevisionModel, CsvRowModel } from "@src/index";
export interface CsvEntryProps {
    onItemsUpdate: (items: ItemRevisionModel[]) => void;
    onBlur: () => void;
}
export interface CsvEntryState {
    csvInputValue?: string;
    csvData: CsvRowModel[];
}
export declare class CsvEntry extends React.Component<CsvEntryProps, CsvEntryState> {
    constructor(props: CsvEntryProps);
    renderHelpButton(): JSX.Element;
    handleCsvChange(event: React.FormEvent<HTMLTextAreaElement>): void;
    handleCsvBlur: () => void;
    render(): JSX.Element;
}
