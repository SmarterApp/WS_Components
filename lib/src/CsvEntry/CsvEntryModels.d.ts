import { ItemRevisionModel } from "../ItemBank/ItemBankModels";
export interface CsvRowModel extends ItemRevisionModel {
    index: number;
}
export declare function parseCsv(csvValue?: string): CsvRowModel[];
