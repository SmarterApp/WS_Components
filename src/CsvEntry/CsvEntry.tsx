import * as React from "react";
import {
  generateTooltip,
  ItemRevisionModel,
  CsvRowModel,
  NamespaceModel,
  parseCsv,
  Accordion
} from "@src/index";

export interface CsvEntryProps {
  namespaces: NamespaceModel[];
  onItemsUpdate: (items: ItemRevisionModel[]) => void;
  onBlur: () => void;
}

export interface CsvEntryState {
  csvInputValue?: string;
  csvData: CsvRowModel[];
}

export class CsvEntry extends React.Component<CsvEntryProps, CsvEntryState> {
  constructor(props: CsvEntryProps) {
    super(props);

    this.state = { csvData: [] };
  }

  renderHelpButton() {
    const helpText: JSX.Element = (
      <p>Enter namespace, bank key, item id, and section for the items you would like to add</p>
    );
    const displayText: JSX.Element = (
      <button
        className="item-nav-btn btn btn-default btn-sm about-item-btn"
        role="button"
        aria-label="Open help text"
      >
        <span className="fa fa-info-circle" aria-hidden="true" />
        Help
      </button>
    );

    return (
      <div className="help-button">
        {generateTooltip({
          helpText,
          displayText,
          displayIcon: undefined
        })}
      </div>
    );
  }

  handleCsvChange(event: React.FormEvent<HTMLTextAreaElement>) {
    const rawCsv: string = event.currentTarget.value;

    this.setState({
      csvInputValue: rawCsv
    });
  }

  handleCsvBlur = () => {
    const { csvInputValue } = this.state;
    const { namespaces } = this.props;
    const csvData = parseCsv(csvInputValue, namespaces);

    this.setState({
      csvData
    });
    this.props.onItemsUpdate(csvData);
    this.props.onBlur();
  };

  render() {
    return (
      <div className="csv-entry-wrapper">
        {this.renderHelpButton()}
        <textarea
          className="csv-text-entry"
          onChange={e => this.handleCsvChange(e)}
          value={this.state.csvInputValue}
          onBlur={this.handleCsvBlur}
        />
      </div>
    );
  }
}
