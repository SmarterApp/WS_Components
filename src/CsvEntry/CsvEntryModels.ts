import { ItemRevisionModel } from "../ItemBank/ItemBankModels";

export interface CsvRowModel extends ItemRevisionModel {
  index: number;
}

export function parseCsv(csvValue: string | undefined) {
  const data: CsvRowModel[] = [];

  if (csvValue) {
    const lines = csvValue.split("\n");
    let index = 0;

    for (const line of lines) {
      const values = line.split(",");

      const row: CsvRowModel = {
        index,
        itemKey: +values[1],
        bankKey: +values[0],
        section: values[2]
      };

      data.push(row);
      index = index + 1;
    }
  }

  return data;
}
