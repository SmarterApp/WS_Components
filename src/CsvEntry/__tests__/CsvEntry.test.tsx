import "jsdom-global/register";
import * as React from "react";
import { shallow, mount } from "enzyme";
import { CsvEntry, CsvEntryProps, CsvEntryState } from "../CsvEntry";
import { ItemRevisionModel } from "../../index";
import { CsvRowModel } from "../CsvEntryModels";

const onItemsUpdateMock = jest.fn();

const item: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "asfe",
  isaap: ""
};

const item2 = { ...item, bankKey: 2332, revision: "asdf" };

const items: ItemRevisionModel[] = [item, item2];

const csvEntryProps: CsvEntryProps = {
  onItemsUpdate: onItemsUpdateMock
};

const csvData1: CsvRowModel = {
  index: 0
};

const csvData2: CsvRowModel = {
  index: 1
};

const csvEntryState: CsvEntryState = {
  csvInputValue: "hello",
  csvData: [csvData1, csvData2]
};

describe("CsvEntry", () => {
  const wrapper = shallow(<CsvEntry {...csvEntryProps} />);

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls handle csv blur", () => {
    const wrapperInstance = wrapper.instance() as CsvEntry;
    wrapper.find(".csv-text-entry").simulate("focus");
    wrapper.find(".csv-text-entry").simulate("blur");
    expect(wrapper).toMatchSnapshot();
  });

  it("calls handle csv change", () => {
    wrapper
      .find(".csv-text-entry")
      .simulate("change", { currentTarget: { value: "hello" } });
    expect(wrapper).toMatchSnapshot();
  });
});
