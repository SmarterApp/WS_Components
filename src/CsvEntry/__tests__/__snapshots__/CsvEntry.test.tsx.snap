// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvEntry calls handle csv blur 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      helpText={
        <p>
          Drag and drop items to add them to the table
        </p>
      }
      position="bottom"
    >
      <button
        aria-label="Open help text"
        className="item-nav-btn btn btn-default btn-sm about-item-btn"
        role="button"
      >
        <span
          aria-hidden="true"
          className="fa fa-info-circle"
        />
        Help
      </button>
    </ToolTip>
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
  />
</div>
`;

exports[`CsvEntry calls handle csv change 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      helpText={
        <p>
          Drag and drop items to add them to the table
        </p>
      }
      position="bottom"
    >
      <button
        aria-label="Open help text"
        className="item-nav-btn btn btn-default btn-sm about-item-btn"
        role="button"
      >
        <span
          aria-hidden="true"
          className="fa fa-info-circle"
        />
        Help
      </button>
    </ToolTip>
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
    value="hello"
  />
</div>
`;

exports[`CsvEntry matches snapshot 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      helpText={
        <p>
          Drag and drop items to add them to the table
        </p>
      }
      position="bottom"
    >
      <button
        aria-label="Open help text"
        className="item-nav-btn btn btn-default btn-sm about-item-btn"
        role="button"
      >
        <span
          aria-hidden="true"
          className="fa fa-info-circle"
        />
        Help
      </button>
    </ToolTip>
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
  />
</div>
`;
